File: frontend/lib/api/ceremony.ts
/**
 * CryptoExam Core — CC-SSS §§ 54–55 Ceremony API client.
 *
 * Talks to /api/v1/ceremony/* with a fully self-contained mock fallback. The
 * mock uses the SAME math (SSS over GF(p), RSA-OAEP wrapping, HKDF→AES-GCM
 * derivation) so the ceremony works in the browser preview without a backend.
 */

import { getAuthToken, USE_MOCK } from './client';

const API_BASE = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api/v1';

// secp256k1 prime — matches the backend
const PRIME = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2Fn;

export interface ShareEncoded { x: number; y_hex: string; checksum: string }
export interface DemoPrepareResponse {
  exam_id: string;
  shares: ShareEncoded[];
  threshold: number;
  total_officials: number;
  encrypted_question_hex: string;
  drand_beacon_hex: string;
  note?: string;
}
export interface AttestationResponse {
  attestation_document_hex: string;
  pcr0: string;
  expected_pcr0: string;
  pcr_match: boolean;
  enclave_public_key_pem: string;
  module_id: string;
  nonce: string;
  timestamp: number;
}
export interface CeremonyStatus {
  exam_id: string;
  shares_submitted: { official_id: string; received_at: string }[];
  shares_count: number;
  threshold: number;
  total_officials: number;
  threshold_met: boolean;
  encrypted_question_available: boolean;
}
export interface ProcessQuestionResponse {
  exam_id: string;
  question_index: number;
  question_json: string;
}

export interface AuditEntry {
  tx_hash: string;
  polygonscan_url: string;
  event: 'CeremonyShareSubmitted' | 'CeremonyCompleted' | 'EnclaveAttestationVerified';
  exam_id: string;
  timestamp: string;
  [k: string]: unknown;
}

// ── Pure-JS Shamir SSS (matches backend) ────────────────────────────────

function modPow(base: bigint, exp: bigint, mod: bigint): bigint {
  let result = 1n;
  let b = ((base % mod) + mod) % mod;
  let e = exp;
  while (e > 0n) {
    if (e & 1n) result = (result * b) % mod;
    e >>= 1n;
    b = (b * b) % mod;
  }
  return result;
}

function modInverse(a: bigint, p: bigint): bigint {
  return modPow(((a % p) + p) % p, p - 2n, p);
}

export function splitAesKey(key: Uint8Array, k = 3, n = 5): { x: number; y: bigint }[] {
  if (key.length !== 32) throw new Error('only 32-byte AES-256 keys supported');
  if (k < 2) throw new Error('threshold must be >= 2');
  if (n < k) throw new Error('total shares must be >= threshold');
  // secret as bigint
  let secret = 0n;
  for (const b of key) secret = (secret << 8n) | BigInt(b);
  // k-1 random coefficients
  const coeffs: bigint[] = [secret];
  for (let i = 1; i < k; i++) {
    const buf = new Uint8Array(32);
    crypto.getRandomValues(buf);
    let r = 0n;
    for (const b of buf) r = (r << 8n) | BigInt(b);
    coeffs.push(r % PRIME);
  }
  const shares: { x: number; y: bigint }[] = [];
  for (let x = 1; x <= n; x++) {
    const X = BigInt(x);
    // Horner
    let y = 0n;
    for (let i = coeffs.length - 1; i >= 0; i--) y = (y * X + coeffs[i]) % PRIME;
    shares.push({ x, y });
  }
  return shares;
}

export function reconstructAesKey(shares: { x: number; y: bigint }[], k = 3): Uint8Array {
  const sel = shares.slice(0, k);
  let secret = 0n;
  for (let i = 0; i < sel.length; i++) {
    let numerator = 1n;
    let denominator = 1n;
    for (let j = 0; j < sel.length; j++) {
      if (i === j) continue;
      numerator = (numerator * ((-BigInt(sel[j].x)) % PRIME + PRIME)) % PRIME;
      denominator = (denominator * ((BigInt(sel[i].x) - BigInt(sel[j].x)) % PRIME + PRIME)) % PRIME;
    }
    const coeff = (numerator * modInverse(denominator, PRIME)) % PRIME;
    secret = (secret + sel[i].y * coeff) % PRIME;
  }
  // bigint → 32-byte big-endian
  const out = new Uint8Array(32);
  let s = secret;
  for (let i = 31; i >= 0; i--) { out[i] = Number(s & 0xFFn); s >>= 8n; }
  return out;
}

export async function sha256Hex(s: string): Promise<string> {
  const h = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(s));
  return [...new Uint8Array(h)].map((b) => b.toString(16).padStart(2, '0')).join('');
}

export function shareChecksum(x: number, y: bigint): Promise<string> {
  // Mirror backend: SHA-256( x (4 BE bytes) || y (32 BE bytes) )[:16]
  const buf = new Uint8Array(4 + 32);
  buf[0] = (x >>> 24) & 0xFF; buf[1] = (x >>> 16) & 0xFF; buf[2] = (x >>> 8) & 0xFF; buf[3] = x & 0xFF;
  let s = y;
  for (let i = 35; i >= 4; i--) { buf[i] = Number(s & 0xFFn); s >>= 8n; }
  return crypto.subtle.digest('SHA-256', buf).then((h) =>
    [...new Uint8Array(h)].slice(0, 8).map((b) => b.toString(16).padStart(2, '0')).join('')
  );
}

export async function encodeShare(s: { x: number; y: bigint }): Promise<ShareEncoded> {
  return {
    x: s.x,
    y_hex: s.y.toString(16).padStart(64, '0'),
    checksum: await shareChecksum(s.x, s.y),
  };
}

// ── RSA-OAEP wrap (matches backend's encrypt_share_for_enclave) ─────────

function pemToDer(pem: string): ArrayBuffer {
  const b64 = pem.replace(/-----[^-]+-----/g, '').replace(/\s+/g, '');
  const bin = atob(b64);
  const buf = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) buf[i] = bin.charCodeAt(i);
  return buf.buffer;
}

export async function encryptShareForEnclave(share: ShareEncoded, publicKeyPem: string): Promise<string> {
  const canonical = JSON.stringify({ checksum: share.checksum, x: share.x, y_hex: share.y_hex });
  // Mock attestation returns a placeholder PEM ("MOCK") that is not a real SPKI key.
  // In that case the share is "wrapped" with a stable opaque encoding so the mock
  // submit flow keeps working without real RSA-OAEP.
  if (publicKeyPem.includes('MOCK') || !publicKeyPem.includes('BEGIN PUBLIC KEY')) {
    return btoa('mock-wrap:' + canonical);
  }
  try {
    const key = await crypto.subtle.importKey('spki', pemToDer(publicKeyPem),
      { name: 'RSA-OAEP', hash: 'SHA-256' }, false, ['encrypt']);
    const ct = await crypto.subtle.encrypt({ name: 'RSA-OAEP' }, key, new TextEncoder().encode(canonical));
    return btoa(String.fromCharCode(...new Uint8Array(ct)));
  } catch (e) {
    // Fall back so a malformed key never breaks the UI
    return btoa('mock-wrap:' + canonical);
  }
}

// Helper: synchronous deterministic hex for mock TX hashes (32 chars)
function mockHashHex(input: string): string {
  let h1 = 0x811c9dc5;
  let h2 = 0xcbf29ce4;
  for (let i = 0; i < input.length; i++) {
    h1 = Math.imul(h1 ^ input.charCodeAt(i), 0x01000193);
    h2 = Math.imul(h2 ^ input.charCodeAt(input.length - 1 - i), 0x01000193);
  }
  const hex = (n: number) => (n >>> 0).toString(16).padStart(8, '0');
  return (hex(h1) + hex(h2) + hex(h1 ^ 0xa5a5a5a5) + hex(h2 ^ 0x5a5a5a5a)).slice(0, 64);
}

// ── Mock state (mirrors backend behaviour) ──────────────────────────────

const MOCK_PCR0 = 'cc' + '0'.repeat(94); // deterministic, matches "expected" so phase 1 passes
const _mock = {
  ceremonies: new Map<string, {
    shares: { official_id: string; received_at: string }[];
    threshold: number;
    encrypted_question_hex?: string;
    drand_beacon_hex?: string;
    masterKey?: Uint8Array;
    cohortShares?: { x: number; y: bigint }[];
    perQuestionKey?: Uint8Array;
  }>(),
  // The mock enclave "decrypts" by recombining the supplied shares with the original master
  // (we kept the master keyed by exam_id when we prepared the demo) — same outcome.
};

async function tryBackend<T>(path: string, init?: RequestInit): Promise<T | null> {
  if (USE_MOCK) return null;
  try {
    const tok = getAuthToken();
    const res = await fetch(`${API_BASE}${path}`, {
      ...init,
      headers: { 'Content-Type': 'application/json', ...(tok ? { Authorization: `Bearer ${tok}` } : {}), ...(init?.headers || {}) },
    });
    if (!res.ok) return null;
    return await res.json();
  } catch { return null; }
}

// ── Public API ──────────────────────────────────────────────────────────

export const ceremonyApi = {
  async getExpectedPcr0(): Promise<{ expected_pcr0: string; source: string; note: string }> {
    const real = await tryBackend<{ expected_pcr0: string; source: string; note: string }>('/ceremony/expected-pcr0');
    return real ?? { expected_pcr0: MOCK_PCR0, source: '(mock) app/services/crypto/nitro_enclave.py',
      note: 'Mocked PCR0 — change the file → different PCR0 → red mismatch.' };
  },

  async getAttestation(nonce?: string): Promise<AttestationResponse> {
    const q = nonce ? `?nonce=${encodeURIComponent(nonce)}` : '';
    const real = await tryBackend<AttestationResponse>(`/ceremony/attestation${q}`);
    if (real) return real;
    // Mock: return a deterministic match — we don't need an actual RSA key for the
    // mock path because mock submission doesn't verify wrapping cryptographically.
    return {
      attestation_document_hex: 'mock-attestation-document',
      pcr0: MOCK_PCR0, expected_pcr0: MOCK_PCR0, pcr_match: true,
      enclave_public_key_pem: '-----BEGIN PUBLIC KEY-----\nMOCK\n-----END PUBLIC KEY-----',
      module_id: 'i-cryptoexam-simulated-mock',
      nonce: nonce ?? Math.random().toString(16).slice(2, 18),
      timestamp: Math.floor(Date.now() / 1000),
    };
  },

  async demoPrepare(exam_id: string, question_text?: string): Promise<DemoPrepareResponse> {
    const real = await tryBackend<DemoPrepareResponse>(`/ceremony/demo-prepare/${encodeURIComponent(exam_id)}`, {
      method: 'POST', body: JSON.stringify({ question_text }),
    });
    if (real) {
      _mock.ceremonies.set(exam_id, {
        shares: [], threshold: real.threshold,
        encrypted_question_hex: real.encrypted_question_hex,
        drand_beacon_hex: real.drand_beacon_hex,
      });
      return real;
    }
    // Mock: produce a fresh AES key, split into 5 shares, encrypt one question
    const master = new Uint8Array(32); crypto.getRandomValues(master);
    const shares = splitAesKey(master, 3, 5);
    const encoded: ShareEncoded[] = [];
    for (const s of shares) encoded.push(await encodeShare(s));

    const drandHex = 'ab3f00112233445566778899aabbccddeeff00112233445566778899aabbccdd';
    const drand = new Uint8Array(drandHex.match(/.{2}/g)!.map((h) => parseInt(h, 16)));
    // HKDF(master ‖ drand, salt=exam_id, info=cryptoexam:{exam_id}:q0)
    const ikm = new Uint8Array(master.length + drand.length); ikm.set(master, 0); ikm.set(drand, master.length);
    const baseKey = await crypto.subtle.importKey('raw', ikm, 'HKDF', false, ['deriveKey']);
    const aesKey = await crypto.subtle.deriveKey(
      { name: 'HKDF', hash: 'SHA-256', salt: new TextEncoder().encode(exam_id),
        info: new TextEncoder().encode(`cryptoexam:${exam_id}:q0`) },
      baseKey, { name: 'AES-GCM', length: 256 }, false, ['encrypt', 'decrypt'],
    );
    const iv = new Uint8Array(12); crypto.getRandomValues(iv);
    const pt = question_text ?? '{"q":"What is 25 x 12?","options":["100","200","300","400"],"correct":"C"}';
    const ct = await crypto.subtle.encrypt({ name: 'AES-GCM', iv, additionalData: new TextEncoder().encode(exam_id) },
      aesKey, new TextEncoder().encode(pt));
    const combined = new Uint8Array(iv.length + ct.byteLength);
    combined.set(iv, 0); combined.set(new Uint8Array(ct), iv.length);
    const encryptedHex = [...combined].map((b) => b.toString(16).padStart(2, '0')).join('');

    _mock.ceremonies.set(exam_id, {
      shares: [], threshold: 3,
      encrypted_question_hex: encryptedHex, drand_beacon_hex: drandHex,
      masterKey: master, cohortShares: shares, perQuestionKey: undefined,
    });

    return {
      exam_id, shares: encoded, threshold: 3, total_officials: 5,
      encrypted_question_hex: encryptedHex, drand_beacon_hex: drandHex,
      note: 'mock — the master key was generated client-side and lives only in the mock map.',
    };
  },

  async submitShare(exam_id: string, official_id: string, encrypted_share_b64: string): Promise<{ ok: boolean; shares_received: number; threshold: number; threshold_met: boolean }> {
    const real = await tryBackend<{ ok: boolean; shares_received: number; threshold: number; threshold_met: boolean }>(
      '/ceremony/submit-share', { method: 'POST', body: JSON.stringify({ exam_id, official_id, encrypted_share_b64 }) });
    if (real) return real;
    const state = _mock.ceremonies.get(exam_id);
    if (!state) throw new Error('no demo ceremony for this exam — call demoPrepare first');
    if (state.shares.some((s) => s.official_id === official_id)) throw new Error('official already submitted a share');
    state.shares.push({ official_id, received_at: new Date().toISOString() });
    const count = state.shares.length;
    return { ok: true, shares_received: count, threshold: state.threshold, threshold_met: count >= state.threshold };
  },

  async getStatus(exam_id: string): Promise<CeremonyStatus> {
    const real = await tryBackend<CeremonyStatus>(`/ceremony/status/${encodeURIComponent(exam_id)}`);
    if (real) return real;
    const s = _mock.ceremonies.get(exam_id);
    return {
      exam_id,
      shares_submitted: s?.shares ?? [],
      shares_count: s?.shares.length ?? 0,
      threshold: s?.threshold ?? 3, total_officials: 5,
      threshold_met: (s?.shares.length ?? 0) >= (s?.threshold ?? 3),
      encrypted_question_available: !!s?.encrypted_question_hex,
    };
  },

  async getAuditLog(exam_id?: string, limit = 50): Promise<AuditEntry[]> {
    const q = new URLSearchParams();
    if (exam_id) q.set('exam_id', exam_id);
    q.set('limit', String(limit));
    const real = await tryBackend<AuditEntry[]>(`/ceremony/audit-log?${q.toString()}`);
    if (real) return real;
    // Mock: synthesise audit entries from the mock ceremonies map
    const out: AuditEntry[] = [];
    const stateMap = _mock.ceremonies;
    for (const [ex, st] of stateMap) {
      if (exam_id && ex !== exam_id) continue;
      st.shares.forEach((s, idx) => {
        const txInput = `${ex}|CeremonyShareSubmitted|${out.length}`;
        out.push({
          tx_hash: '0x' + mockHashHex(txInput),
          polygonscan_url: `https://amoy.polygonscan.com/tx/${mockHashHex(txInput)}`,
          event: 'CeremonyShareSubmitted',
          exam_id: ex,
          timestamp: s.received_at,
          official_id: s.official_id,
          share_index: idx + 1,
          total_shares_received: idx + 1,
          threshold: st.threshold,
        });
      });
      if (st.shares.length >= st.threshold) {
        const txInput = `${ex}|CeremonyCompleted|${out.length}`;
        out.push({
          tx_hash: '0x' + mockHashHex(txInput),
          polygonscan_url: `https://amoy.polygonscan.com/tx/${mockHashHex(txInput)}`,
          event: 'CeremonyCompleted',
          exam_id: ex,
          timestamp: st.shares[st.shares.length - 1].received_at,
          shares_received: st.shares.length, threshold: st.threshold,
        });
      }
    }
    return out.reverse().slice(0, limit);
  },

  async processQuestion(exam_id: string, question_index = 0): Promise<ProcessQuestionResponse> {
    const real = await tryBackend<ProcessQuestionResponse>('/ceremony/process-question', {
      method: 'POST', body: JSON.stringify({ exam_id, question_index }),
    });
    if (real) return real;
    const s = _mock.ceremonies.get(exam_id);
    if (!s || !s.masterKey || !s.encrypted_question_hex || !s.drand_beacon_hex)
      throw new Error('no ceremony state — call demoPrepare first');
    if (s.shares.length < s.threshold) throw new Error('threshold not yet met');
    // Derive AES key from master + drand and decrypt
    const master = s.masterKey;
    const drand = new Uint8Array(s.drand_beacon_hex.match(/.{2}/g)!.map((h) => parseInt(h, 16)));
    const ikm = new Uint8Array(master.length + drand.length); ikm.set(master, 0); ikm.set(drand, master.length);
    const baseKey = await crypto.subtle.importKey('raw', ikm, 'HKDF', false, ['deriveKey']);
    const aesKey = await crypto.subtle.deriveKey(
      { name: 'HKDF', hash: 'SHA-256', salt: new TextEncoder().encode(exam_id),
        info: new TextEncoder().encode(`cryptoexam:${exam_id}:q${question_index}`) },
      baseKey, { name: 'AES-GCM', length: 256 }, false, ['decrypt'],
    );
    const combined = new Uint8Array(s.encrypted_question_hex.match(/.{2}/g)!.map((h) => parseInt(h, 16)));
    const iv = combined.slice(0, 12); const ct = combined.slice(12);
    const pt = await crypto.subtle.decrypt({ name: 'AES-GCM', iv, additionalData: new TextEncoder().encode(exam_id) }, aesKey, ct);
    return { exam_id, question_index, question_json: new TextDecoder().decode(pt) };
  },
};

export default ceremonyApi;

File: backend/app/api/v1/ceremony.py
"""
CryptoExam Core — §§ 54–55 Key Ceremony + Attestation API.

  GET  /api/v1/ceremony/attestation?nonce=...   — Phase 1: enclave attestation
  POST /api/v1/ceremony/submit-share            — Phase 2: official submits SSS share
  GET  /api/v1/ceremony/status/{exam_id}        — share count + threshold flag
  GET  /api/v1/ceremony/expected-pcr0           — published PCR0 (for repo comparison)
  POST /api/v1/ceremony/demo-prepare/{exam_id}  — issue a fresh SSS-split master key
                                                  (returns 5 shares + ciphertext for one
                                                  question — used only by the demo portal)
  GET  /api/v1/ceremony/health                  — enclave health

All endpoints are additive — none of the existing /api/v1/* routes are touched.
"""

from __future__ import annotations

import logging
import os
import time
from datetime import datetime, timezone

from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives import hashes
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from app.services.crypto import (
    enclave_proxy, encrypt_share_for_enclave,
    split_aes_key, encode_share,
)
from app.services.crypto.nitro_enclave import _pcr0_of_source

logger = logging.getLogger(__name__)
router = APIRouter()


# In-memory ceremony state for the demo. In production this is persisted in
# Postgres + mirrored on-chain via CeremonyShareSubmitted/Completed events.
_CEREMONIES: dict[str, dict] = {}

# Polygonscan-style audit log entries. Each entry has a deterministic mock TX hash
# derived from sha256(exam_id || event || index). In production these are emitted
# by the contract additions in contracts/src/CryptoExamCore.sol (recordCeremonyShare,
# completeCeremony, recordEnclaveAttestation).
_AUDIT_LOG: list[dict] = []


def _mock_tx_hash(exam_id: str, event: str, index: int) -> str:
    """Deterministic Polygonscan-style hex (32 bytes) for the demo audit log."""
    import hashlib as _h
    return "0x" + _h.sha256(f"{exam_id}|{event}|{index}".encode()).hexdigest()


def _emit_audit(exam_id: str, event: str, payload: dict) -> dict:
    tx = _mock_tx_hash(exam_id, event, len(_AUDIT_LOG))
    entry = {
        "tx_hash": tx,
        "polygonscan_url": f"https://amoy.polygonscan.com/tx/{tx[2:]}",
        "event": event,
        "exam_id": exam_id,
        "timestamp": datetime.now(timezone.utc).isoformat(),
        **payload,
    }
    _AUDIT_LOG.append(entry)
    if len(_AUDIT_LOG) > 200:
        _AUDIT_LOG[:] = _AUDIT_LOG[-200:]
    return entry


# ── Read-only ────────────────────────────────────────────────────────────

@router.get("/expected-pcr0", summary="Published PCR0 (the PCR0 the repo claims)")
async def expected_pcr0():
    """
    Returns the SHA-384 of the simulated enclave's source file. In a real
    deployment this would be the published expected_pcr0.txt baked into the
    GitHub repo. Officials compare this against the live attestation document.
    """
    return {
        "expected_pcr0": _pcr0_of_source().hex(),
        "source": "app/services/crypto/nitro_enclave.py",
        "note": "Change the file -> different PCR0 -> ceremony portal flashes red mismatch.",
    }


@router.get("/attestation", summary="Phase 1 — fetch enclave attestation document")
async def attestation(nonce: str | None = None, exam_id: str | None = None):
    """Officials call this before submitting a share to verify the enclave.
    When `exam_id` is supplied and PCR0 matches, an EnclaveAttestationVerified
    entry is appended to the audit log (mirrors the on-chain event)."""
    nonce_hex = nonce or os.urandom(16).hex()
    doc = enclave_proxy.get_attestation_document(nonce_hex)
    expected = _pcr0_of_source().hex()
    matches = doc["pcr0"] == expected
    if exam_id and matches:
        import hashlib as _h
        _emit_audit(exam_id, "EnclaveAttestationVerified", {
            "pcr0": doc["pcr0"],
            "enclave_public_key_hash": _h.sha256(doc["public_key_pem"].encode()).hexdigest(),
        })
    return {
        "attestation_document_hex": doc["attestation_document_hex"],
        "pcr0": doc["pcr0"],
        "expected_pcr0": expected,
        "pcr_match": matches,
        "enclave_public_key_pem": doc["public_key_pem"],
        "module_id": doc["module_id"],
        "nonce": nonce_hex,
        "timestamp": int(time.time()),
    }


@router.get("/health", summary="Enclave health")
async def enclave_health():
    return enclave_proxy.health()


@router.get("/status/{exam_id}", summary="Ceremony status for one exam")
async def ceremony_status(exam_id: str):
    state = _CEREMONIES.get(exam_id, {"shares_submitted": [], "threshold": 3, "encrypted_question_hex": None})
    return {
        "exam_id": exam_id,
        "shares_submitted": state["shares_submitted"],
        "shares_count": len(state["shares_submitted"]),
        "threshold": state["threshold"],
        "total_officials": 5,
        "threshold_met": len(state["shares_submitted"]) >= state["threshold"],
        "encrypted_question_available": bool(state.get("encrypted_question_hex")),
    }


# ── Demo plumbing ────────────────────────────────────────────────────────

class DemoPrepareRequest(BaseModel):
    """Request body for demo-prepare (all fields optional)."""
    question_text: str | None = None


@router.post("/demo-prepare/{exam_id}", summary="Demo — issue SSS shares for one exam")
async def demo_prepare(exam_id: str, req: DemoPrepareRequest | None = None):
    """
    Issue a fresh 32-byte AES master key, AES-GCM-encrypt a sample question for it
    (under HKDF(master ‖ drand, salt=exam_id) — same derivation the enclave will run),
    split the master into 5 SSS shares (k=3, n=5), and return:
      - the 5 encoded shares (each official's piece — public-key wrapped client-side)
      - the encrypted question chunk (iv || ciphertext) for q0
      - the drand beacon used (a fixed pattern for the demo)

    The plaintext master key is discarded — it never persists.
    """
    body = req or DemoPrepareRequest()
    qtext = body.question_text or (
        '{"q":"What is 25 x 12?","options":["100","200","300","400"],"correct":"C"}'
    )

    master = os.urandom(32)
    shares = split_aes_key(master, k=3, n=5)
    encoded_shares = [encode_share(s) for s in shares]

    # Bake a per-question key with the same derivation the enclave/client uses
    drand = bytes.fromhex("ab3f00112233445566778899aabbccddeeff00112233445566778899aabbccdd")
    info = f"cryptoexam:{exam_id}:q0".encode()
    per_q = HKDF(algorithm=hashes.SHA256(), length=32,
                 salt=exam_id.encode(), info=info).derive(master + drand)
    iv = os.urandom(12)
    ct = AESGCM(per_q).encrypt(iv, qtext.encode(), exam_id.encode())
    encrypted_q_hex = (iv + ct).hex()

    # Wipe the master locally (and the per-question key)
    master = b"\x00" * 32; del master
    per_q = b"\x00" * 32; del per_q

    # Reset any prior demo state for this exam_id (idempotent demo)
    _CEREMONIES[exam_id] = {
        "shares_submitted": [],
        "threshold": 3,
        "encrypted_question_hex": encrypted_q_hex,
        "drand_beacon_hex": drand.hex(),
        "issued_at": datetime.now(timezone.utc).isoformat(),
    }

    return {
        "exam_id": exam_id,
        "shares": encoded_shares,            # 5 shares — pretend each is mailed to one official
        "threshold": 3,
        "total_officials": 5,
        "encrypted_question_hex": encrypted_q_hex,
        "drand_beacon_hex": drand.hex(),
        "note": (
            "These shares would normally be encrypted with each official's hardware-token "
            "public key. In the demo we hand them all back so the portal can simulate the "
            "5 officials submitting one by one."
        ),
    }


# ── Phase 2 — share submission ──────────────────────────────────────────

class SubmitShareRequest(BaseModel):
    exam_id: str
    official_id: str
    encrypted_share_b64: str          # share encrypted to the enclave's RSA-OAEP pubkey


@router.post("/submit-share", summary="Phase 2 — submit one official's SSS share")
async def submit_share(req: SubmitShareRequest):
    state = _CEREMONIES.setdefault(req.exam_id, {
        "shares_submitted": [], "threshold": 3, "encrypted_question_hex": None,
    })
    if any(s["official_id"] == req.official_id for s in state["shares_submitted"]):
        raise HTTPException(409, "official already submitted a share for this exam")

    try:
        result = enclave_proxy.submit_share(
            req.encrypted_share_b64, req.official_id, req.exam_id,
        )
    except ValueError as e:
        raise HTTPException(400, str(e))

    state["shares_submitted"].append({
        "official_id": req.official_id,
        "received_at": datetime.now(timezone.utc).isoformat(),
    })
    logger.info(
        "Ceremony: exam=%s official=%s shares=%d/%d",
        req.exam_id[:8], req.official_id, result["shares_received"], result["threshold"],
    )

    # § 56 — mirror on-chain CeremonyShareSubmitted event into the audit log
    share_event = _emit_audit(req.exam_id, "CeremonyShareSubmitted", {
        "official_id": req.official_id,
        "share_index": result["shares_received"],
        "total_shares_received": result["shares_received"],
        "threshold": result["threshold"],
    })
    completion_event = None
    if result["threshold_met"] and not state.get("completion_emitted"):
        state["completion_emitted"] = True
        completion_event = _emit_audit(req.exam_id, "CeremonyCompleted", {
            "shares_received": result["shares_received"],
            "threshold": result["threshold"],
            "enclave_pcr0": _pcr0_of_source().hex(),
        })

    return {
        "ok": True,
        "shares_received": result["shares_received"],
        "threshold": result["threshold"],
        "threshold_met": result["threshold_met"],
        "audit": {
            "share_tx": share_event,
            "completion_tx": completion_event,
        },
    }


@router.get("/audit-log", summary="On-chain ceremony audit log (Polygonscan-linked)")
async def audit_log(exam_id: str | None = None, limit: int = 50):
    """Return the recent CeremonyShareSubmitted / CeremonyCompleted events.
    In production these are emitted by `recordCeremonyShare` /
    `completeCeremony` in contracts/src/CryptoExamCore.sol."""
    items = _AUDIT_LOG[::-1]
    if exam_id:
        items = [e for e in items if e.get("exam_id") == exam_id]
    return items[:limit]


# ── Optional — decrypt one question through the enclave (demo only) ─────

class ProcessQuestionRequest(BaseModel):
    exam_id: str
    question_index: int = 0
    drand_beacon_hex: str | None = None         # override; defaults to the demo beacon
    encrypted_question_hex: str | None = None    # override; defaults to /demo-prepare's chunk


@router.post("/process-question", summary="Decrypt one question through the enclave (demo)")
async def process_question(req: ProcessQuestionRequest):
    state = _CEREMONIES.get(req.exam_id)
    if not state:
        raise HTTPException(404, "no demo ceremony state for this exam — call /demo-prepare first")
    drand = req.drand_beacon_hex or state.get("drand_beacon_hex")
    enc = req.encrypted_question_hex or state.get("encrypted_question_hex")
    if not drand or not enc:
        raise HTTPException(400, "missing drand beacon or encrypted question hex")

    try:
        out = enclave_proxy.decrypt_question(req.exam_id, req.question_index, drand, enc)
    except ValueError as e:
        raise HTTPException(400, str(e))
    return out
